### DISCUSSION TITLE
Trying to build a strace-like output for capturing file I/O activity

### DISCUSSION THREAD

**Discussion by rodrigonascimento** [bpftrace/bpftrace#2717] [Q&A] [answered]
Hi, 

I just started learning bpftrace programming and I decided to create a script to trace file I/O syscalls. I thought it would be a good experiment since I can validate it against a strace output. :) 

I'm starting with only a few syscalls at the moment like open*, read, and close, but I also want to include lseek, fsync, fdatasync, pread64, write, and pwrite64.

Since I'm a noob when it comes to bpftrace programming, I would like to check w/ someone more experienced if the logic I'm applying here makes sense. 

I start my script by gathering which PID I want bpftrace to act on:
```
BEGIN 
{

    @pidtrace = $1 
}
```
My understanding is that to do anything with a file (read it or write it) we need to open the file first, and as a result the OS will assign a file descriptor to it.  Thus, I have a block tracing t:syscalls:sys_enter_open, filtering it by PID and since this looks like the only place from where I can grab the filename, I have a map by TID storing filenames:
```
tracepoint:syscalls:sys_enter_open,
tracepoint:syscalls:sys_enter_openat,
tracepoint:syscalls:sys_enter_openat2
/pid == @pidtrace/
{
    @filename[tid] = args->filename;
}
```
Then, I have a block tracing t:syscalls:sys_exit_open filtering it by @filename[tid], to collect the file descriptor(FD) which I will use to create another map called fdfilename to keep track of the relationship TID, FD, filename. If we get a fd greater than or equal to 0 we successfully opened the file and I add the filename to the @fdfilename[tid,$fd] map. Then, I delete the filename stored in the @filename[tid] map.
```
tracepoint:syscalls:sys_exit_open*
/ @filename[tid] / 
{
    $ret = args->ret;
    $fd = $ret >= 0 ? $ret : -1;
    if ( $fd >= 0 ) {
        printf("%ld syscall=%d probe=%s process=%s pid=%d tid=%d filename=%s fd=%ld\n", nsecs, args->__syscall_nr, probe, comm, pid, tid, str(@filename[tid]), $fd);
        @fdfilename[tid,$fd] = @filename[tid];
    }
    delete(@filename[tid]);
}
```
Next, I have a block tracing t:syscalls:sys_enter_read filtering it by TID and FD via the @fdfilename[tid, args->fd] map. Here I want to keep track of FD and how many bytes is being requested.
```
tracepoint:syscalls:sys_enter_read
/ @fdfilename[tid,args->fd] /
{
    @op_read_fd[tid] = args->fd;
    @op_read_req_size_b[tid] = args->count;
}
```
Then, I have a block tracing t:syscalls:sys_exit_read filtering it by TID and FD via the @fdfilename[tid, @op_read_fd[tid]] map. I captured the number of bytes read, and delete the corresponding data from the @op_read_fd and @op_read_req_size_b maps.
```
tracepoint:syscalls:sys_exit_read
/ @fdfilename[tid,@op_read_fd[tid]] /
{
    $ret = args->ret;
    $bytes_read = $ret >0 ? $ret : -1;
    
    printf("%ld syscall=%d probe=%s process=%s pid=%d tid=%d filename=%s fd=%ld req_size_bytes=%ld bytes_read=%ld\n", 
        nsecs, args->__syscall_nr, probe, comm, pid, tid, str(@fdfilename[tid,@op_read_fd[tid]]), @op_read_fd[tid], @op_read_req_size_b[tid], 
        $bytes_read);
    
    delete(@op_read_fd[tid]);
    delete(@op_read_req_size_b[tid]);
}
```
Next, I have a block tracing t:syscalls:sys_enter_close filtering it by TID and FD via the @fdfilename[tid, args->fd] map. I print some info and collect the FD for the file getting closed in the @op_close[tid] map.
```
tracepoint:syscalls:sys_enter_close
/ @fdfilename[tid,args->fd] /
{
    printf("%ld syscall=%d probe=%s process=%s pid=%d tid=%d filename=%s fd=%ld\n", nsecs, args->__syscall_nr, probe, comm, pid, tid, str(@fdfilename[tid,args->fd]), args->fd);
    @op_close[tid] = args->fd;
}
```
Finally, I have a block tracing t:syscalls:sys_exit_close filtered by TID and FID via the @fdfilename[tid, @op_close[tid]] map. Since we don't need to keep track of the filename anymore, I delete the entry from the @fdfilename map.
```
tracepoint:syscalls:sys_exit_close
/ @fdfilename[tid, @op_close[tid]] /
{
    delete(@fdfilename[tid, @op_close[tid]]);
}
```
The last block clear all the maps used in the script:
```
END
{
    clear(@pidtrace);
    clear(@filename);
    clear(@op_read_fd);
    clear(@op_read_req_size_b);
    clear(@op_pread64_fd);
    clear(@op_pread64_req_size_b);
    clear(@op_pread64_offset);
    clear(@op_close);
    clear(@fdfilename);
}
```
Does that make sense? Any thoughts/comments?

**Comment by fbs** [comment, marked as answer]
looks quite ok to me :)

**Reply by rodrigonascimento** [reply]
thanks for taking the time to look at it, @fbs . I really appreciate it.

**Comment by grissomsh** [comment]
Connecting the dots ... 

```
#!/usr/bin/env bpftrace

#include <linux/fs.h>

BEGIN 
{
    @pidtrace = (uint64)$1 
}

tracepoint:syscalls:sys_enter_open, tracepoint:syscalls:sys_enter_openat / pid == @pidtrace /
{
    @filename[tid] = args->filename;
}

tracepoint:syscalls:sys_exit_open, tracepoint:syscalls:sys_exit_openat / @filename[tid] / 
{
    $ret = args->ret;
    $fd = $ret >= 0 ? $ret : -1;
    if ( $fd >= 0 ) {
        printf("%s %-3d %-40s %-20s %-8d %-8d %ld %s\n", strftime("%H:%M:%S:%f", nsecs), args->nr, probe, comm, pid, tid, $fd, str(@filename[tid]));
        @fdfilename[tid,$fd] = @filename[tid];
    }
    delete(@filename[tid]);
}

tracepoint:syscalls:sys_enter_read / @fdfilename[tid,args->fd] /
{
    @op_read_fd[tid] = args->fd;
    @op_read_req_size_b[tid] = args->count;
}

tracepoint:syscalls:sys_exit_read / @fdfilename[tid,@op_read_fd[tid]] /
{
    $ret = args->ret;
    $bytes_read = $ret >0 ? $ret : -1;
    
    printf("%s %-3d %-40s %-20s %-8d %-8d %ld %ld %ld %s\n", strftime("%H:%M:%S:%f", nsecs), args->nr, probe, comm, pid, tid, @op_read_req_size_b[tid], $bytes_read, @op_read_fd[tid], str(@fdfilename[tid,@op_read_fd[tid]]));
    
    delete(@op_read_fd[tid]);
    delete(@op_read_req_size_b[tid]);
}

tracepoint:syscalls:sys_enter_close / @fdfilename[tid,args->fd] /
{
    printf("%s %-3d %-40s %-20s %-8d %-8d %ld %s\n", strftime("%H:%M:%S:%f", nsecs), args->nr, probe, comm, pid, tid, args->fd, str(@fdfilename[tid,args->fd]));
    @op_close[tid] = args->fd;
}

tracepoint:syscalls:sys_exit_close / @fdfilename[tid, @op_close[tid]] /
{
    delete(@fdfilename[tid, @op_close[tid]]);
}

END
{
    clear(@pidtrace);
    clear(@filename);
    clear(@op_read_fd);
    clear(@op_read_req_size_b);
    clear(@op_close);
    clear(@fdfilename);
}
```
